### DISCUSSION TITLE
Is there a way to use AsyncReadBody and indicate the total content length?

### DISCUSSION THREAD

**Discussion by t-rapp** [tokio-rs/axum#2074] [Q&A] [answered]
In some of the responses returned from my axum-http server application I have a virtual data source implementing `tokio::io::AsyncRead` returning audio and video media data. When I wrap that source with `axum_extra::body::AsyncReadBody` the HTTP response seems to use chunked transfer encoding without a total content-length header. Within client applications (media players) this leads to an unknown total length of the file. Thus seeking in the client is either disabled or leads to wrong results until the full file is transferred.

Is there a way to use AsyncReadBody without chunked transfer encoding and including the `Content-Length` header? Will award bonus points if there is a way to support byte range requests on top of that :)

**Comment by davidpdrsn** [comment, marked as answer]
Nothing built in (though maybe there should be). Your best option is to implement your own [`http_body::Body`](https://docs.rs/http-body/latest/http_body/trait.Body.html) type that wraps your async read body and overrides `Body::size_hint`

**Reply by t-rapp** [reply]
Thanks for the pointers. Will take a deeper look at these types.

**Reply by t-rapp** [reply]
Implemented my own version of [`axum_extra::body::AsyncReadBody`](https://docs.rs/axum-extra/0.7.4/axum_extra/body/struct.AsyncReadBody.html) with tracking of remaining bytes inspired by the implementation of [`http_body::Limited`](https://docs.rs/http-body/0.4.5/http_body/struct.Limited.html). My version of `size_hint()` returns `SizeHint::with_exact(remaining)` and the content-length header is part of the response now.
